src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferDictionaryContainsMethods.cs

```diff
@@ -0,0 +1,208 @@
+
+                //  Once we have an applicable Contains method and an applicable property reference, we search for an applicable ContainsXXX
+                //  method on the IDictionary`2 receiver. 
+
+                //  A ContainsXXX method is applicable if:
+                //      1) It has a boolean return type
+                //      2) It is publically accessible
+                //      3) Its name is "ContainsKey" if property is "Keys", or "ContainsValue" if property is "Values". 
+                //      4) It has exactly one parameter of the correct type:
+                //          - If the method is a "ContainsKey" method, its type must be TKey. 
+                //          - If the method is a "ContainsValue" method, its type must be TValue.
+
+                var invocation = (IInvocationOperation)context.Operation;
+                IMethodSymbol containsMethod = invocation.TargetMethod;
+
+                if (containsMethod.Name != ContainsMethodName
+                    || containsMethod.ReturnType.SpecialType != SpecialType.System_Boolean
+                    || !TryGetPropertyReferenceOperation(invocation, out IPropertySymbol? property)
+                    || !TryGetConstructedDictionaryType(property.ContainingType, out INamedTypeSymbol? constructedDictionaryType))
+                {
+                    return;
+                }
+
+                //  At this point, we know that the method being invoked is a method called "Contains" that has a boolean return type.
+                //  We also know that the method is being invoked on a property belonging to a type that implements IDictionary`2. We will
+                //  compare the types used to construct IDictionary`2 to the parameter type of the Contains method. 
+
+                ITypeSymbol keyType = constructedDictionaryType.TypeArguments[0];
+                ITypeSymbol valueType = constructedDictionaryType.TypeArguments[1];
+
+                //  We use Parameters.Last() because the first argument could be the key/value collection, depending
+                //  on whether the method is an extension method and whether the language is C# or Visual Basic. 
+
+                ITypeSymbol containsParameterType = containsMethod.Parameters[^1].Type;
+
+                if (property.Name == KeysPropertyName)
+                {
+                    if (!containsParameterType.Equals(keyType, SymbolEqualityComparer.Default))
+                        return;
+
+                    //  Now we search for an accessible ContainsKey method that returns boolean and accepts a single
+                    //  parameter of the type that was substituted for TKey.
+
+                    IMethodSymbol? containsKeyMethod = property.ContainingType.GetMembers(ContainsKeyMethodName)
+                        .OfType<IMethodSymbol>()
+                        .WhereAsArray(x => x.ReturnType.SpecialType == SpecialType.System_Boolean &&
```

Comment:
Why `WhereAsArray` followed by `FirstOrDefault`?

Reply:
Oops, I'll fix that.